src/webgpu/web_platform/canvas/getCurrentTexture.spec.ts

```diff
@@ -260,3 +265,80 @@ g.test('resize')
     currentTexture = ctx.getCurrentTexture();
     t.expect(prevTexture === currentTexture);
   });
+
+g.test('expiry')
+  .desc(
+    `
+Test automatic WebGPU canvas texture expiry on all canvas types with the following requirements:
+- getCurrentTexture returns the same texture object within the same frame, throughout:
+  - after previous frame update the rendering
+  - before current frame update the rendering
+  - in a microtask off the current frame task
+- getCurrentTexture returns a new texture object and the old texture object becomes invalid as soon as possible after HTML update the rendering.
+  `
+  )
+  .params(u =>
+    u //
+      .combine('canvasType', kAllCanvasTypes)
+      .combine('prevFrameCallsite', ['requestPostAnimationFrame', 'requestAnimationFrame'] as const)
+  )
+  .fn(t => {
+    const { canvasType, prevFrameCallsite } = t.params;
+    const ctx = t.initCanvasContext(t.params.canvasType);
+    // Create a bindGroupLayout to test invalid texture view usage later.
+    const bgl = t.device.createBindGroupLayout({
+      entries: [
+        {
+          binding: 0,
+          visibility: GPUShaderStage.COMPUTE,
+          texture: {},
+        },
+      ],
+    });
+
+    // The fn is called immediately after previous frame updating the rendering.
+    // Polyfill by calling the callback by setTimeout, in the requestAnimationFrame callback (for onscreen canvas)
+    // or after transferToImageBitmap (for offscreen canvas).
+    function requestPostAnimationFrame(fn: () => void) {
+      t.requestNewFrameOnCanvasType(canvasType, ctx, () => {
+        timeout(fn);
+      });
+    }
+
+    function checkGetCurrentTexture() {
+      // Call getCurrentTexture on previous frame.
+      const prevTexture = ctx.getCurrentTexture();
+
+      // Call getCurrentTexture immediately after the frame, the texture object should stay the same.
+      queueMicrotask(() => {
+        t.expect(prevTexture === ctx.getCurrentTexture());
+      });
+
+      // Call getCurrentTexture immediately after this frame updating the rendering.
+      // It should return a new texture object.
+      timeout(() => {
```

Comment:
This is the earliest requirement we want for chrome implementation. The spec allows expiry happens immediately before rendering. https://gpuweb.github.io/gpuweb/#-automatic-expiry-task-source. So we may want to chagne this to a warn, or remove this check and replace with a less strict check.

Reply:
My understanding of the spec is there are 3 cases

1. next task (eg, setTimeout, rAF)
2. rAF
3. update the rendering

3 is only covered if the "update the rendering" step is run and it is *NOT* run in response to rAF. It is run if something in the page changed. In the test I wrote I added (and removed next rAF) and 1x1 pixel div (display: fixed, left: 0, top:0, width: 1px, height 1px, opacity 50%) to try to cover that case. Of course it can only do this in a page, not in a worker.